src/json.hpp

```diff
@@ -2756,9 +2705,7 @@ class lexer
     /// return string value
     const std::string get_string()
     {
-        // yytext cannot be returned as char*, because it may contain a null
-        // byte (parsed as "\u0000")
-        return std::string(yytext.data(), yylen);
+        return std::move( yytext );
```

Comment:
This could lead to nasty bugsif called from a lvalue, could you provide lvalue and rvalue overloads?

Reply:
I would have to look up the caller, but maybe we can also return a const reference.

Reply:
We can return a const ref for lvalues and move for rvalues. I remember someone on reddit discussing the rvalue support in the Library, which could be far better.

But that's not an emergency at all

Reply:
I don't think so; but, on review -- I noticed that the get_string function was (incorrectly) returning a 'const std::string', preventing any move re-use of the returned std::string anyway!  Removing the const appears to have resulted in (another) pretty dramatic speed-up.  New push forthcoming...

Reply:
Well simply calling `get_string` twice would become problematic. It adds more boilerplate to add those overloads, but the code will always be correct. We should use `std::move(lexer).get_string()` to be explicit about this side effect

Reply:
I think consuming from a lexer (as in: moving data out of the lexer) is a common pattern. We can add documentation though.

Reply:
This get_string method is being used for keys and values; both of which are ultimately moved into a std::map key, or into a BasicJsonType value; there are no lvalue users.

Reply:
Yup, appears to improve performance by close to 10% on some benchmarks.  Nice.

Reply:
Yes, calling get_string twice would (surprisingly) result in an empty string.  Perhaps renaming it to `move_string()` would result in less surprising behavior?  I'm not sure I understand your `std::move(lexer).get_string()` recommendation.

Reply:
I get your point. Let's hope nobody will get bitten by that then :)

Reply:
I haven't seen your last answer sorry. You can overload on lvalues and rvalues:

```cpp
class lexer {
std::string const& get_string() & {
  return _content;
}

std::string get_string() && {
  return std::move(_content);
}
};
```

The first function can only be called on lvalues, while the latter can only be called on rvalues.
Hence the `std::move(lexer).get_string()`, which will call the second overload, while `lexer.get_string()` will call the first one.

I've used this pattern on classes holding huge data, and I would call the rvalue overload at the very end of the processing, while calling the lvalue one during processing.

But a `move_string` is fine with me too.